src/concurrency/thread.rs

```diff
@@ -1070,10 +1070,26 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
         thread.state = ThreadState::Terminated;
 
         let current_span = this.machine.current_span();
-        for ptr in
-            this.machine.threads.thread_terminated(this.machine.data_race.as_mut(), current_span)
-        {
-            this.deallocate_ptr(ptr.into(), None, MiriMemoryKind::Tls.into())?;
+        let thread_local_allocations =
+            this.machine.threads.thread_terminated(this.machine.data_race.as_mut(), current_span);
+        match this.get_active_thread() {
+            // For main thread, add thread-local statics to static roots and skip deallocating
+            // backing memory
+            ThreadId(0) =>
```

Comment:
Instead of matching here, it makes more sense to change the code here:

https://github.com/rust-lang/miri/blob/8bbc93ca31f29b10168919f241d9c9be6b69f77e/src/eval.rs#L250-L252

the main thread is already special, so instead of calling to common code and then special-casing again -- just don't call to common code. :)

Reply:
I tried doing this, but doing so causes many tests to fail since we do still want to run some of the other logic in `terminate_active_thread()` (e.g. waiting on other threads to finish):
> Error: the main thread terminated without waiting for all remaining threads

Reply:
Ah, fair. We  probably need a flag to be passed to `terminate_active_thread` then indicating whether the thread-locals should be freed or leaked.

Reply:
Added a `leak_tls` flag